libraries/rush-lib/src/logic/snapshots/InputSnapshot.ts

```diff
@@ -0,0 +1,437 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import { createHash, type Hash } from 'crypto';
+import ignore, { type Ignore } from 'ignore';
+
+import { Path, Sort } from '@rushstack/node-core-library';
+
+import type { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import type { IOperationSettings, RushProjectConfiguration } from '../../api/RushProjectConfiguration';
+import { type IReadonlyLookupByPath, LookupByPath } from '../LookupByPath';
+import { RushConstants } from '../RushConstants';
+
+/**
+ * @beta
+ */
+export type IRushConfigurationProjectForSnapshot = Pick<
+  RushConfigurationProject,
+  'projectFolder' | 'projectRelativeFolder'
+>;
+
+/**
+ * @internal
+ */
+export interface IRushSnapshotProjectMetadata {
+  /**
+   * The contents of rush-project.json for the project, if available
+   */
+  projectConfig?: RushProjectConfiguration;
+  /**
+   * A map of operation name to additional files that should be included in the hash for that operation.
+   */
+  additionalFilesByOperationName?: ReadonlyMap<string, ReadonlySet<string>>;
+}
+
+interface IInternalRushSnapshotProjectMetadata {
+  /**
+   * The contents of rush-project.json for the project, if available
+   */
+  projectConfig?: RushProjectConfiguration;
+  /**
+   * Cached filter of files that are not ignored by the project's `incrementalBuildIgnoredGlobs`.
+   * @param filePath - The path to the file to check
+   * @returns true if the file path is an input to all operations in the project, false otherwise
+   */
+  projectFilePathFilter?: (filePath: string) => boolean;
+  /**
+   * The cached Git hashes for all files in the project folder.
+   */
+  hashes: Map<string, string>;
+  /**
+   * The list of additional file paths that are applicable to each operation name, if any.
+   */
+  additionalFilesByOperationName?: ReadonlyMap<string, ReadonlySet<string>>;
```

Comment:
```suggestion
interface IInternalRushSnapshotProjectMetadata extends IRushSnapshotProjectMetadata {
  /**
   * Cached filter of files that are not ignored by the project's `incrementalBuildIgnoredGlobs`.
   * @param filePath - The path to the file to check
   * @returns true if the file path is an input to all operations in the project, false otherwise
   */
  projectFilePathFilter?: (filePath: string) => boolean;
  /**
   * The cached Git hashes for all files in the project folder.
   */
  hashes: Map<string, string>;
```